### ISSUE TITLE
Binding URI parameters to Noda Time types with Web API/MVC

### ISSUE THREAD

**Issue by zaccharles** [nodatime/nodatime.serialization#3]
This issue is to track future development of https://github.com/nodatime/nodatime/pull/368 in this new repo.

**Comment by zaccharles** [comment]
@jskeet What's the idea in this repo regarding 1.3.x and 2.0?

**Comment by jskeet** [comment]
This repo will be serialisation targeting 2.0 and higher... Within the next week, I hope to release 2.0 and update the code in here to refer to that... All other code should do likewise.

**Comment by TheBeardedLlama** [comment]
hey guys, is there any progress on this? or a draft I could work from?
or something I could help with?

**Comment by jskeet** [comment]
I know nothing about model binding basically - hopefully @zaccharles can help you :)

**Comment by zaccharles** [comment]
I haven't had (made) time to work on this and I'm not sure when I will. Given how much time has passed, it would be quite a lot of effort for me to get back up to speed on all this as I don't use it anymore. 

@TheBeardedLlama I suggest you look at nodatime/nodatime#368 if you need something in the meantime.

**Comment by TheBeardedLlama** [comment]
kk

I need to get some of this done for something I'm working on now

Shall I look to create a PR with the result or are you not interested in this functionality?

**Comment by jskeet** [comment]
I think I'd like a clear explanation of what the functionality *is*, aside from anything else. That will help to decide where it lives - probably in this repo, but in a new project, or an existing one? What else does it tie in with, etc?

**Comment by TheBeardedLlama** [comment]
Fair enough

This functionality would be to wire up NodaTime in an IModelBinder, so that web api/mvc core 2.0 can deal with the data types natively. This is specifically when they are used as part of model input to an action.
The json.net code in this project deals with serialisation when the return type contains NodaTime data types.

**Comment by jskeet** [comment]
Okay, so if it's ASP.NET-specific it sounds like it shouldn't be in NodaTime.Serialization.JsonNet, but it might *depend* on that? Can it easily target both ASP.NET and ASP.NET Core, or would that require two separate projects?

**Comment by TheBeardedLlama** [comment]
I believe you are correct. (I was just following the trail of tickets hence the question)

I don't think it needs access to json.net stuff. I'm using the core Pattern<T> stuff.

As for compatibility, there seem to be quite a few differences in model binding between core 1 and 2 and I'm assuming standard ASP.net.
I would probably say you need to keep them separated (pun intended).
Unless someone really smart can make it all work from one place.

**Comment by TheBeardedLlama** [comment]
so what are your thoughts @jskeet ?

I did most of the plumbing yesterday, so there wouldn't be much left to do to make this available for others if you're interested
I appreciate the conundrum regarding placement though

**Comment by jskeet** [comment]
If it doesn't require the Json.NET code, that's good. It sounds like it may be worth having separate ASP.NET and ASP.NET Core projects - hopefully we can get away without targeting ASP.NET Core 1.0 and ASP.NET Core 2.0 differently.

If you could put together a PR for one or both of those, that would be very helpful. Work and personal life are both exceptionally busy right now, so I don't expect to get time to review it in the near future unless it's really small, I'm afraid - it's likely to be at least a couple of weeks before I can dedicate significant time.

**Comment by TheBeardedLlama** [comment]
Ah don't worry about lack of time, same here :)

So are we talking about a project(s) in this repo?

**Comment by jskeet** [comment]
Assuming it feels serialization-based, this repo would be most appropriate, yes.

**Comment by TheBeardedLlama** [comment]
Ok, I'll make a start and we can take it from there as time permits

**Comment by stevenguh** [comment]
I have not tested this extensively. But it seems like adding `TypeConverter` on the NodaTime type will make `ModelBinder` unnecessary. I tried this with an `Instant` type as a parameter on an action of a controller in WebApi, the parameter seems to bind correctly with the `TypeConverter`.
The use of `TypeConverter` is also mentioned in https://docs.microsoft.com/en-us/aspnet/web-api/overview/formats-and-model-binding/parameter-binding-in-aspnet-web-api#type-converters

See https://github.com/nodatime/nodatime.serialization/issues/2 for `TypeConverter` discussion

**Comment by jskeet** [comment]
@stevenguh: Does that also work if instead of decorating the original types with a TypeConverter, they're annotated using TypeDescriptor, as per #2? Ideally I'd like to separate the serialization concerns as far as possible. (I couldn't for XML and binary serialization, but that was a necessity.)

**Comment by TheBeardedLlama** [comment]
@jskeet indeed, but that does not apply to core (I don't think)

**Comment by TheBeardedLlama** [comment]
@jskeet doh! very last comment on https://docs.microsoft.com/en-us/aspnet/core/mvc/advanced/custom-model-binding is about typeconverter
such an idiot!
apologies for wasting both our time
I will look at that option. Typeconverter will deal with the conundrum around versions as well btw
I will update once I'm done

**Comment by TheBeardedLlama** [comment]
yeah so `TypeConverter` works a treat :flushed: 

Also it's easy to add using `TypeDescriptor`

However, as you pointed out in one of the linked issues, I'm not sure what effect that would have in a XAML/design environment

**Comment by stevenguh** [comment]
@jskeet Yes, I believe so. From a runtime perspective, I think annotating `TypeConverter` statically at compile time or via `TypeConverter` will be pretty much the same. So yes, I think we can separate the serialization concerns without overloading the main repo with a bunch of `TypeConverter` annotations.

**Comment by jskeet** [comment]
So am I right in saying that this one change would address this issue, #2 and #12? Woot. Anyone else want to do this, or should I look into it myself?

**Comment by TheBeardedLlama** [comment]
It sounds like @stevenguh has done most of the work

I've also done it now to replace the model binders solution. Not for all types mind you. However, I could do it given some design guidelines

**Comment by stevenguh** [comment]
I basically have all the TypeConverters for NodaTime implemented except for `Interval` I can try to submit a PR this weekend.

**Comment by stevenguh** [comment]
I created a fork, but there are some issues
https://github.com/stevenguh/nodatime.serialization/tree/typeconverter

- Putting TypeConverters inside of the NodaTime.Serialization.Json or just in this repo feels a bit weird
- .NET Standard 1.6 doesn't have some API needed like Attributes property in MemeberInfo (It's available in .NET standard 2.0)

**Comment by jskeet** [comment]
> I created a fork, but there are some issues
https://github.com/stevenguh/nodatime.serialization/tree/typeconverter

One quick drive-by comment: NodaConveters needs to be NodaConverters :)

> - Putting TypeConverters inside of the NodaTime.Serialization.Json or just in this repo feels a bit weird

Still not having really got my head round this, what would you suggest instead? Adding them in the main project? (That has one downside of requiring whatever our next main project release is, rather than just being able to target 2.0.)

One vague issue I have around this is around time zone providers - if the attributes are attached directly in Noda Time, the time zone provider can't be specified by the user. I really need to see how this all works to know how much of a problem this is.

> - .NET Standard 1.6 doesn't have some API needed like Attributes property in MemeberInfo (It's available in .NET standard 2.0)

I'd expect everything we need to be there, possibly just with different names - is GetCustomAttributes what you're after here? (I can't see an Attributes property so I'm not sure... and my internet connection is too poor to stomach too much research right now.)

**Comment by stevenguh** [comment]
I fixed the typo :)

I mean the putting `TypeConverter` which is not specific to Json or serialization, in the namespace of `NodaTime.Serialization` or `NodaTime.Seriallization.Json` feels a bit weird. Maybe all the `TypeConverter` stuff should be in `NodaTime.Serialization.TypeConverter`. Just to be clear, I'm not suggesting putting `TypeConverter` in the main repo.

Yes, basically it means that a user can only have one zone provider through out an application. If not, then a user will have to keep overriding the zone provider attribute by using `TypeDescriptor`.

`typeof(T).GetTypeInfo().GetCustomAttributes()` doesn't returns attributes annotated with `TypeDescriptor`.

Furthermore, in order to use TypeConverter in .NET standard 1.0+, we have to update the dependency to Json.NET 10.0.3. See https://github.com/JamesNK/Newtonsoft.Json/commit/2e7b24e9b1873a615bbc8ae2302fd3d174cdb9f2

To sum up the issues:
- The work around for getting the zone provider for `ZonedDateTime`, `DateTimeZone` by using `DateTimeZoneProviderAttribute` annotated on the same type with `TypeDescriptor` requires the API in .NET standard 2.0
- Json.NET needs to be at least 10.0.3 in order for it to pick up the `TypeConverter` correctly in .NET standard. (The Type Converters I had in https://github.com/stevenguh/nodatime.serialization/commit/b8b332772ababd047351a9eac5677ef1bd51a493 should work fine in .NET framework 4.5+ with Json.NET 6+)
- The TypeConverters are generic enough that they might not belong in `NodaTime.Serialization.Json` namespace. (`NodaTime.Serialization.TypeConverter` maybe ok, which means a new project?)

**Comment by stevenguh** [comment]
I updated my fork to use .NET standard 2.0 (.NET Core 2.0 in the test and benchmark), and Json.NET 10.0.3. However, if we roll with that, we will be requiring people who use this package to use framework that's compatible with .NET Standard 2.0 and Json.NET 10.0.3. I am not sure what else we could do here.

I also discover some quirks when using TypeConverter:
- The sequence to call `TypeDescriptor.AddAttribute` and `JsonConverter` for value is important (basically we will be using TypeConverter in replace of the JsonConverter)
- The sequence to call `TypeDescriptor.AddAttribute` and `JsonConvert` for dictionary key is NOT important.

If we are using `TypeConverter` in replace of `JsonConverter` we have to call `TypeDescriptor.AddAttribute` before actually calling any serialization method in Json.NET. I think Json.NET cached result of if any TypeConverter on the first serialization of that type.

I guess if we going forward with this with using TypeCovnerter for everything, then adding all type converters has to be the first step of an app, and we will also be assuming there will only be one zone provider. To be specific, one can change the zone provider by change zone in `DateTimeZoneProviderAttribute` by calling `TypeDescriptor.AddAttribute` again. However, after Json.NET cached the type converter, one cannot change the type converter.

The code to demo the quirks...
```csharp
Instant now = SystemClock.Instance.GetCurrentInstant();
ZonedDateTime zdt = now.InZone(DateTimeZoneProviders.Tzdb["America/Los_Angeles"]);
Dictionary<ZonedDateTime, ZonedDateTime> zdtDict = new Dictionary<ZonedDateTime, ZonedDateTime> { [zdt] = zdt };

// Before any TypeConverter.
// Comment this part to try out the sequence
string zdtJsonBefore = JsonConvert.SerializeObject(zdt);
string zdtDictJsonBefore = JsonConvert.SerializeObject(zdtDict);

// Set up TypeConverters
TypeDescriptor.AddAttributes(typeof(ZonedDateTime), new DateTimeZoneProviderAttribute(DateTimeZoneProviders.Tzdb));
TypeDescriptor.AddAttributes(typeof(ZonedDateTime), new TypeConverterAttribute(typeof(ZonedDateTimeTypeConverter)));

string zdtJson = JsonConvert.SerializeObject(zdt);
string zdtDictJson = JsonConvert.SerializeObject(zdtDict);

JsonSerializerSettings s = new JsonSerializerSettings().ConfigureForNodaTime(DateTimeZoneProviders.Tzdb);
string zdtJsonWithSettings = JsonConvert.SerializeObject(zdt, s);
string zdtDictJsonWithSettings = JsonConvert.SerializeObject(zdtDict, s);
```

**Comment by jskeet** [comment]
Thanks for all of that. I don't want to start requiring netstandard2.0 yet... I'd really like to find a netstandard1.3 solution if I possibly can. I can't look in detail now, but will do when I get a chance. Thanks so much for all your research so far.

**Comment by buvinghausen** [comment]
https://github.com/nodatime/nodatime/pull/1237

I think you can close this out with the caveat that you either need to pull the code from 1237 or wait for the 3.0 release which will contain the TypeConverters.

**Comment by jskeet** [comment]
Hooray, thanks for pointing that out. Admittedly I *think* that means we won't be able to use custom formats, but I doubt that'll be an issue. Closing.
